src/mscorlib/corefx/System/Security/SecureString.Unix.cs

```diff
@@ -0,0 +1,295 @@
+                {
+                    _buffer.ReleasePointer();
+                }
+            }
+        }
+
+        private void SetAtCore(int index, char c)
+        {
+            // Overwrite the character at the specified index
+            _buffer.Write((ulong)(index * sizeof(char)), c);
+        }
+
+        internal unsafe IntPtr MarshalToStringCore(bool globalAlloc, bool unicode)
+        {
+            int length = _decryptedLength;
+
+            byte* bufferPtr = null;
+            IntPtr stringPtr = IntPtr.Zero, result = IntPtr.Zero;
+            try
+            {
+                _buffer.AcquirePointer(ref bufferPtr);
+                if (unicode)
+                {
+                    int resultLength = (length + 1) * sizeof(char);
+                    stringPtr = globalAlloc ? Marshal.AllocHGlobal(resultLength) : Marshal.AllocCoTaskMem(resultLength);
+                    Buffer.MemoryCopy(
+                        source: bufferPtr,
+                        destination: (byte*)stringPtr.ToPointer(),
+                        destinationSizeInBytes: resultLength,
+                        sourceBytesToCopy: length * sizeof(char));
+                    *(length + (char*)stringPtr) = '\0';
+                }
+                else
+                {
+                    int resultLength = Encoding.UTF8.GetByteCount((char*)bufferPtr, length) + 1;
+                    stringPtr = globalAlloc ? Marshal.AllocHGlobal(resultLength) : Marshal.AllocCoTaskMem(resultLength);
+                    int encodedLength = Encoding.UTF8.GetBytes((char*)bufferPtr, length, (byte*)stringPtr, resultLength);
+                    Debug.Assert(encodedLength + 1 == resultLength, $"Expected encoded length to match result, got {encodedLength} != {resultLength}");
+                    *(resultLength - 1 + (byte*)stringPtr) = 0;
+                }
+
+                result = stringPtr;
+            }
+            finally
+            {
+                // If there was a failure, such that result isn't initialized, 
+                // release the string if we had one.
+                if (stringPtr != IntPtr.Zero && result == IntPtr.Zero)
+                {
+                    UnmanagedBuffer.ZeroMemory((byte*)stringPtr, (ulong)(length * sizeof(char)));
+                    MarshalFree(stringPtr, globalAlloc);
+                }
+
+                if (bufferPtr != null)
+                {
+                    _buffer.ReleasePointer();
+                }
+            }
+
+            return result;
+        }
+
+        // -----------------------------
+        // ---- PAL layer ends here ----
+        // -----------------------------
+
+        private void EnsureCapacity(int capacity)
+        {
+            // Make sure the requested capacity doesn't exceed SecureString's defined limit
+            if (capacity > MaxLength)
+            {
+                throw new ArgumentOutOfRangeException("capacity", SR.ArgumentOutOfRange_Capacity);
```

Comment:
Nit: `nameof(capacity)`? (I see there are some other uses of `nameof` in this change, so presumably this language feature can be used from corelib.)